### DISCUSSION TITLE
Record Linkage on two imported csv files

### DISCUSSION THREAD

**Discussion by kodeplex** [moj-analytical-services/splink#1559] [General]
I am quite new in python and tryingto use/learn Splink with documentation/step by step tutorial. I have tried to adapt the following tutorial on linking "Linking two tables of persons"
- [Link](https://moj-analytical-services.github.io/splink/demos/examples/duckdb/link_only.html)

Since the example is based on a single file and splitting, I modified the sample code to import the splitted files as csv. Then the code is not working, not sure why it's failing.

I have the temp Codelab Notebook here: https://colab.research.google.com/drive/1NAPABJtaItPZhxs1RRMjN6D97YbiSU0m?usp=sharing

`from duckdb import read_csv
from splink.datasets import splink_datasets
df = splink_datasets.fake_1000

df_l = df.sample(frac=0.5)
df_r = df.drop(df_l.index)
#df_l.head(2)

(df_l).to_csv ("/content/drive/MyDrive/SplinkFiles/df_l.csv")
(df_r).to_csv("/content/drive/MyDrive/SplinkFiles/df_r.csv")

df_l=read_csv("/content/drive/MyDrive/SplinkFiles/df_l.csv")
df_r=read_csv("/content/drive/MyDrive/SplinkFiles/df_r.csv")`

`
from splink.duckdb.linker import DuckDBLinker
from splink.duckdb.blocking_rule_library import block_on
import splink.duckdb.comparison_library as cl
import splink.duckdb.comparison_template_library as ctl

settings = {
    "link_type": "link_only",
    "blocking_rules_to_generate_predictions": [
        block_on("first_name"),
        block_on("surname"),
    ],
    "comparisons": [
        ctl.name_comparison("first_name",),
        ctl.name_comparison("surname"),
        ctl.date_comparison("dob", cast_strings_to_date=True),
        cl.exact_match("city", term_frequency_adjustments=True),
        ctl.email_comparison("email", include_username_fuzzy_level=False),
    ],
}

linker = DuckDBLinker([df_l, df_r], settings, input_table_aliases=["df_left", "df_right"])
deterministic_rules = [
    "l.first_name = r.first_name and levenshtein(r.dob, l.dob) &lt;= 1",
    "l.surname = r.surname and levenshtein(r.dob, l.dob) &lt;= 1",
    "l.first_name = r.first_name and levenshtein(r.surname, l.surname) &lt;= 2",
    "l.email = r.email"
]

linker.estimate_probability_two_random_records_match(deterministic_rules, recall=0.7)
linker.estimate_u_using_random_sampling(max_pairs=1e6, seed=1)
results = linker.predict(threshold_match_probability=0.9)
results.as_pandas_dataframe(limit=5)`

**The error seems to be here:
linker = DuckDBLinker([df_l, df_r], settings, input_table_aliases=["df_left", "df_right"])**

**Comment by RobinL** [comment]
What's the error message?

**Reply by kodeplex** [reply]
I am getting the following error, I have not touched the this part of the code.

---------------------------------------------------------------------------
ParserException                           Traceback (most recent call last)
[<ipython-input-29-3c36877d657b>](https://localhost:8080/#) in <cell line: 24>()
     22 
     23 ###############################
---> 24 linker = DuckDBLinker([df_l, df_r], settings, input_table_aliases=["df_left", "df_right"])
     25 deterministic_rules = [
     26     "l.first_name = r.first_name and levenshtein(r.dob, l.dob) &lt;= 1",

13 frames
[/usr/local/lib/python3.10/dist-packages/splink/duckdb/linker.py](https://localhost:8080/#) in as_record_dict(self, limit)
     47             sql += f" limit {limit}"
     48 
---> 49         return self.linker._con.query(sql).to_df().to_dict(orient="records")
     50 
     51     def as_pandas_dataframe(self, limit=None):

ParserException: Parser Error: syntax error at or near "column0"
LINE 1: ...────────┬───────────┬────────────┬──────────┬────────────┬────────────────────...

**Reply by RobinL** [reply]
The error message `ParserException: Parser Error: syntax error at or near "column0"` seems to be arising from the change you made to read the CSV files directly.

One common cause of this error is that when you write and then read the CSV files, the index of the DataFrame might be saved as an additional column (usually named "Unnamed: 0" or similar) and then read back in as a regular column. When you pass this DataFrame with the additional column to Splink, it may cause the SQL query to break.

Here's a way to potentially fix the issue by explicitly setting the `index=False` when saving the DataFrames to CSV:

```python
(df_l).to_csv("/content/drive/MyDrive/SplinkFiles/df_l.csv", index=False)
(df_r).to_csv("/content/drive/MyDrive/SplinkFiles/df_r.csv", index=False)
```

You can also try to print out the DataFrames `df_l` and `df_r` after reading them from the CSV files to make sure they have the correct structure and do not include any unexpected columns.

**Comment by Mahaktirole** [comment]
Hi, I am using splink to compare two CSV file using link_only as link_type in settings. It is making duplicate records. For example if we have two IDs , unique I'd = 1 and unique I'd =2 present in both CSV files and pass the blocking rule criteria for comparison with each other then splink is is generating both combinations of this pair like (1,2) and (2,1). How can I stop splink to compare (2,1) when (1,2) is already compared?

**Reply by RobinL** [reply]
Could you provide a reproducible example of the problem?  These should be filtered out automatically by Splink, so if you're seeing duplicates then it may be a bug

**Comment by Mahaktirole** [comment]
input_df_1
![image](https://github.com/moj-analytical-services/splink/assets/104124647/4e16cf7f-bb36-49c5-9956-e13194f969bf)  
input_df_2
![image](https://github.com/moj-analytical-services/splink/assets/104124647/4c7b4303-bd12-47cd-b588-c97e8b3f590c)
Resulting pairs
![image](https://github.com/moj-analytical-services/splink/assets/104124647/26342ffd-34ab-4b9e-8db5-ad3721d582cb)

I am trying to apply 'link_only' between input_df_1 and input_df_2 and getting resulting pairs which are to be compared.
I am using splink==2.1.0rc1, unique id is entity_id column and code is below.

linker = Splink(settings, [input_df_1, input_df_2], spark)
comparisons = linker.get_scored_comparisons()

Note: Please don't get confused by not seeing match key, match probability and other expected columns. I have deleted them and added few new column by some other logic but pairs of entity_ids are generated using splink only and I can see duplicates there. for example - pair (1,2) and (2,1) both are present.

**Comment by Mahaktirole** [comment]
input_df_1

|entity_id|first_name|last_name|dob|city|email||
0|0|Robert|9999|24-Jun-1971||robert255@smith.net||
1|1|Robert|Allen|24-May-1971||roberta25@smith.net||
2|2|9999|Allen|24-Jun-1971|London|roberta25@smith.net||

input_df_2
|entity_id|first_name|last_name|dob|city|email
0|0|Robert|Alan|24-Jun-1971||robert255@smith.net
1|1|Robert|Allen|24-May-1971||roberta25@smith.net
2|2|Rob|Allen|24-Jun-1971|London|roberta25@smith.net
3|3|Robert|Alen|24-Jun-1971|Lonon|
4|4|Grace||26-Apr-1997|Hull|grace.kelly52@jones.com
5|5|Grace|Kelly|26-Apr-1991||grace.kelly52@jones.com
6|6|Logan|pMurphy|01-Aug-1973||
7|7|||03-AUg-2015|Portsmouth|evied56@harris-bailey.net

**Reply by RobinL** [reply]
(1,2)
and 
(2,1)
are different comparisons

because (1,2) is comparing id 1 in input df 1 to id 2 in input df 2
and
because (2,1) is comparing id 2 in input df 1 to id 1 in input df 2

**Comment by Mahaktirole** [comment]
ohh okay , Thankyou so much. 
Also, Is there any way to not create pairs for comparing two records having same unique_id in two separate files? 

example, if I have a record with unique_id=1 in input_df_1 and again I have a record with same unique_id=1 in input_df_2. Splink is comparing these two records considering them different records as they are present in different input files. Is there any way to avoid comparing them and only compare records with different unique_ids?

**Reply by Mahaktirole** [reply]
Can someone please update?

**Reply by RobinL** [reply]
In your blocking rules you could include something like:
```
"blocking_rules_to_generate_predictions": ["l.first_name = r.first_name and l.unique_id < r.unique_id"],
```
